io_gripper_controller/src/io_gripper_controller.cpp

```diff
@@ -0,0 +1,1376 @@
+      {
+        check_state_ios_ = false;
+        for (const auto & high_val : state_params.high)
+        {
+          setResult = find_and_get_state(high_val, state_value_);
+          if (!setResult)
+          {
+            RCLCPP_ERROR(
+              get_node()->get_logger(), "Failed to get the state for %s", high_val.c_str());
+          }
+          else
+          {
+            if (abs(state_value_ - 1.0) < std::numeric_limits<double>::epsilon())
+            {
+              check_state_ios_ = true;
+            }
+            else
+            {
+              check_state_ios_ = false;
+              break;
+            }
+          }
+        }
+        for (const auto & low_val : state_params.low)
+        {
+          setResult = find_and_get_state(low_val, state_value_);
+          if (!setResult)
+          {
+            RCLCPP_ERROR(
+              get_node()->get_logger(), "Failed to get the state for %s", low_val.c_str());
+          }
+          else
+          {
+            if (abs(state_value_ - 0.0) < std::numeric_limits<double>::epsilon())
+            {
+              check_state_ios_ = true;
+            }
+            else
+            {
+              check_state_ios_ = false;
+              break;
+            }
+          }
+        }
+        if (check_state_ios_)
+        {
+          closed_state_name_ = state_name;
+          gripper_state_buffer_.writeFromNonRT(gripper_state_type::SET_AFTER_COMMAND);
+          break;
+        }
+      }
+      break;
+    case gripper_state_type::SET_AFTER_COMMAND:
+      closed_state_values_ = params_.close.state.possible_closed_states_map.at(closed_state_name_);
+
+      for (const auto & high_val : closed_state_values_.set_after_command_high)
+      {
+        setResult = find_and_set_command(high_val, 1.0);
+        if (!setResult)
+        {
+          RCLCPP_ERROR(
+            get_node()->get_logger(), "Failed to set the command state for %s", high_val.c_str());
+        }
+      }
+
+      for (const auto & low_val : closed_state_values_.set_after_command_low)
+      {
+        setResult = find_and_set_command(low_val, 0.0);
+        if (!setResult)
+        {
+          RCLCPP_ERROR(
+            get_node()->get_logger(), "Failed to set the command state for %s", low_val.c_str());
+        }
+      }
+      for (size_t i = 0; i < params_.close.state.possible_closed_states_map.at(closed_state_name_)
+                               .joint_states.size();
+           ++i)
+      {
+        joint_state_values_[i] =
+          params_.close.state.possible_closed_states_map.at(closed_state_name_).joint_states[i];
+      }
+      gripper_state_buffer_.writeFromNonRT(gripper_state_type::IDLE);
+      closeFlag_.store(false);
+      gripper_service_buffer_.writeFromNonRT(service_mode_type::IDLE);
+      break;
+    default:
+      break;
+  }
+}
+
+void IOGripperController::handle_gripper_state_transition_open(const gripper_state_type & state)
```

Comment:
I find generally this function somewhat inconsistent with the closing state checks. Can we unify this a bit. Especially the part when setting low and high states. I would except there more or less copy-paste code, or better yet calling subfunctions as we constantly do the same.